cmd/manila-csi-plugin/main.go

```diff
@@ -17,38 +17,87 @@ limitations under the License.
 package main
 
 import (
-	goflag "flag"
-	"k8s.io/cloud-provider-openstack/pkg/csi/manila/manilaclient"
+	"flag"
+	"fmt"
+	"os"
 
-	flag "github.com/spf13/pflag"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
 	"k8s.io/cloud-provider-openstack/pkg/csi/manila"
+	"k8s.io/cloud-provider-openstack/pkg/csi/manila/manilaclient"
+	"k8s.io/component-base/logs"
 	"k8s.io/klog"
 )
 
 var (
-	endpoint      = flag.String("endpoint", "unix://tmp/csi.sock", "CSI endpoint")
-	driverName    = flag.String("drivername", "manila.csi.openstack.org", "name of the driver")
-	nodeID        = flag.String("nodeid", "", "this node's ID")
-	protoSelector = flag.String("share-protocol-selector", "", "specifies which Manila share protocol to use")
-	fwdEndpoint   = flag.String("fwdendpoint", "", "CSI Node Plugin endpoint to which all Node Service RPCs are forwarded. Must be able to handle the file-system specified in share-protocol-selector")
+	endpoint      string
+	driverName    string
+	nodeID        string
+	protoSelector string
+	fwdEndpoint   string
+	userAgentData []string
 )
 
 func main() {
-	klog.InitFlags(nil)
-	if err := goflag.Set("logtostderr", "true"); err != nil {
-		klog.Exitf("failed to set logtostderr flag: %v", err)
+	flag.CommandLine.Parse([]string{})
+
+	cmd := &cobra.Command{
+		Use:   os.Args[0],
+		Short: "CSI Manila driver",
+		PersistentPreRun: func(cmd *cobra.Command, args []string) {
+			// Glog requires this otherwise it complains.
+			flag.CommandLine.Parse(nil)
+
+			// This is a temporary hack to enable proper logging until upstream dependencies
+			// are migrated to fully utilize klog instead of glog.
+			klogFlags := flag.NewFlagSet("klog", flag.ExitOnError)
+			klog.InitFlags(klogFlags)
+
+			// Sync the glog and klog flags.
+			cmd.Flags().VisitAll(func(f1 *pflag.Flag) {
+				f2 := klogFlags.Lookup(f1.Name)
+				if f2 != nil {
+					value := f1.Value.String()
+					f2.Value.Set(value)
+				}
+			})
+		},
+		Run: func(cmd *cobra.Command, args []string) {
+			d, err := manila.NewDriver(nodeID, driverName, endpoint, fwdEndpoint, protoSelector, &manilaclient.ClientBuilder{UserAgentData: userAgentData})
+			if err != nil {
+				klog.Fatalf("driver initialization failed: %v", err)
+			}
+
+			d.Run()
+		},
 	}
 
-	manilaClientBuilder := &manilaclient.ClientBuilder{}
+	cmd.Flags().AddGoFlagSet(flag.CommandLine)
+
+	cmd.PersistentFlags().StringVar(&endpoint, "endpoint", "unix://tmp/csi.sock", "CSI endpoint")
+	cmd.MarkPersistentFlagRequired("endpoint")
+
+	cmd.PersistentFlags().StringVar(&driverName, "drivername", "manila.csi.openstack.org", "name of the driver")
+	cmd.MarkPersistentFlagRequired("drivername")
+
+	cmd.PersistentFlags().StringVar(&nodeID, "nodeid", "", "this node's ID")
+	cmd.MarkPersistentFlagRequired("nodeid")
+
+	cmd.PersistentFlags().StringVar(&protoSelector, "share-protocol-selector", "", "specifies which Manila share protocol to use")
```

Comment:
same here a default protocol??

Reply:
Protocol selector heavily depends on how Manila is configured and deployed and this in turn dictates what goes into `--fwdendpoint`...IMHO there's too many possibilities, making any assumptions on what should be the default would probably end up being wrong anyway. I think the safest thing to do here is to make the users explicitly set which share protocol they're interested in.

Reply:
I agree that we shouldn't set a default protocol selector, i.e. we should force the deployer to choose.  That requires some thought up front on the part of the deployer but will in the end be much less confusing than doing an apparently successful deployment and then discovering at time of PVC fulfillment that it does not work as intended.

Reply:
got it, but can we specify some valid values here?? and have a validity check here only??

Reply:
@adisky fixed in https://github.com/kubernetes/cloud-provider-openstack/pull/716/commits/aaeb55134a5a67aa8b66f992ec77d191183d82fb

Also, I'm not too familiar with Cobra, but apparently even if a flag has a default value, it must not be marked as required and left unspecified - otherwise it complains that a required flag is missing (which, on second though, makes sense) => I've removed `MarkPersistentFlagRequired` for flags that have default values specified.